File: 剑指offer/3，从尾到头打印链表.py
"""
从尾到头遍历一遍列表
"""


class Node(object):
    def __init__(self, val):
        self.val = val
        self.next = None


def print_list_from_tail_to_head(head):
    if head is None:
        return []

    cur = head
    temp = []
    while cur is not None:
        temp.append(cur.val)
        cur = cur.next
    rev_temp = temp[::-1]
    return rev_temp


a = Node(1)
b = Node(2)
c = Node(3)
d = Node(4)
a.next = b
b.next = c
c.next = d


res = print_list_from_tail_to_head(a)
print(res)

File: cpp_version/剑指offer/3，从尾到头打印链表.cpp
//
// Created by Jiang,Xinyang on 2020/7/27.
//

#include <iostream>
#include <vector>

using namespace std;


struct ListNode {
    int val;
    struct ListNode *next;

    ListNode(int x) :
            val(x), next(NULL) {
    }
};


vector<int> printListFromTailToHead(ListNode* head) {
    vector<int> result;
    if (head == nullptr)
        return result;

    ListNode* cur = head;
    while(cur != nullptr){
        result.insert(result.begin(), cur->val);
        cur = cur->next;
    }

    return result;
}


int main() {

    ListNode* a = new ListNode(1);
    ListNode* b = new ListNode(2);
    ListNode* c = new ListNode(3);
    a->next = b;
    b->next = c;

    vector<int> res = printListFromTailToHead(a);

    for (int x : res)
        cout << x << " ";
    cout << endl;

    delete a;
    delete b;
    delete c;
}